### PR TITLE
Validator V2 Updates - v2.03

### PR DESCRIPTION
**Pull request by nosvalds**
## Summary
- Updates mapping.xml to include validator rule metadata
- Creates codelist_rules.json for use in validator V2
- Workflow to auto-update ruleset .csv repository
- Workflow to push changes to redis instance

## Details
- basic mapping to codelist_rules complete
- pulling in allowed codes
- update format to handle more cases
- handle conditions for linked vocab to codelist
- add validation rules metadata to mapping
- generate with updated non-emb codelists
- update with additional UNSDG mappings
- change severity from danger to error
- export codelist validator rules to csv format
- change id due to duplicate rule
- update and format
- fix lint and add workflow to push to redis
- fix linting issues
- add job to trigger csv update on rule tracker repo, remove csv files
- remove quote from message, remove all csv references
- add un system function to sector vocab

### CODE REVIEW

**Comment by nosvalds** [comment]
Testing out the IATI-Reference-Generator with this branch to ensure it doesn't break anything in the website SSOT updates now.

**Review by akmiller01 — APPROVED** [review]
No impact on the IATI-Reference-Generator from my local tests!

**Comment by andylolz** [comment]
Hi @nosvalds,

I currently use mapping.xml in other projects, and I’m concerned this change looks like it could be breaking… I think this mapping file (and more generally, the contents of this repo) should be considered part of the Single Source of Truth, so care should be taken when making changes.

In this instance, it looks like these changes relate to the IATI Validator? If that’s the case, would it be possible to make these validator-related changes outside of the SSOT? This is probably also good practice in terms of separation of concerns. If the changes relate to the IATI Validator, perhaps they can live in one of the IATI Validator repos?

Thanks,

**Comment by nosvalds** [comment]
Hi @andylolz, can you clarify what is breaking? 

The change to mapping.xml just added a new subelement of `<validation-rules>` for `<mapping>` which includes metadata about the related validation rule. 

The existing `mapping/path`, `mapping/codelist/@ref`, `mapping/condition` are unchanged.

**Comment by andylolz** [comment]
> can you clarify what is breaking

My concern was that this change _could be_ breaking. It wasn’t immediately easy to tell, because it adds ~20k lines of code that are extraneous to the SSOT. It looks like both you and @akmiller01 wanted to check first that it didn’t cause problems for IATI-Reference-Generator [[1]](https://github.com/IATI/IATI-Codelists/pull/257#issuecomment-865115129) [[2]](https://github.com/IATI/IATI-Codelists/pull/257#pullrequestreview-688606398), so probably not unreasonable for me to have similar concerns about other infrastructure.

I notice a [sample of the mapping.xml is provided in the README for this repo](https://github.com/IATI/IATI-Codelists#codelist-mapping), and hasn’t been updated as part of this PR. The fact that care hasn’t been taken to update the documentation also made me a bit nervous.

It would be great if any changes to the version branches of the SSOT repos could go through some sort of management process (similar to [the codelist management process](https://iatistandard.org/en/iati-standard/upgrades/how-we-manage-the-standard/codelist-management/)) just so users don’t have to panic any time there’s a PR in these SSOT repos, that it a) could be merged at any moment, and b) could potentially be breaking. E.g. a guaranteed one week grace period before changes are merged, in order to allow time to review changes and raise any objections.

> The change to mapping.xml just added a new subelement of `<validation-rules>` for `<mapping>` which includes metadata about the related validation rule.

These `validation-rules` don’t have any meaning outside of the IATI Validator – these priorities, IDs, categories are not part of the standard. So it’s unclear to me why it belongs in an SSOT repo. It’s not part of the standard, it’s part of some separate tool that also happens to be managed by the secretariat. It’s also going to be a pain for you to make changes to them (e.g. to amend the message text, or add translations etc) if they’re managed in this versioned repo.

File: .gitignore
Comment (andylolz) [inline .gitignore:6]: I wonder if this line is a typo, and the plan here was in fact to gitignore codelist_rules.json?

**Comment by andylolz** [comment]
It looks like codelist_rules.json contains non-embedded codelist codes, so will presumably need to be updated every time a non-embedded codelist is changed? I wonder if the intention was instead to gitignore codelist_rules.json, and that this ~650kb file wasn’t intended to be included in this PR?

**Comment by andylolz** [comment]
It looks like codelist_rules.json contains non-embedded codelists (up to fourteen copies in some cases), so will presumably need to be updated every time a non-embedded codelist is changed?

I wonder if the intention was instead to gitignore codelist_rules.json, and that this ~650kb file wasn’t intended to be included in this PR?

**Comment by stevieflow** [comment]
> These validation-rules don’t have any meaning outside of the IATI Validator – these priorities, IDs, categories are not part of the standard. So it’s unclear to me why it belongs in an SSOT repo. 

I would be *very * concerned if this is the case.   The standard cannot be changed or updated in this way.  The standard has to be managed and maintained independently of any tools that use it.  There has been a careful balance of this since IATI day one.

**Comment by stevieflow** [comment]
@nosvalds any chance to receive some feedback on the posts ^^ ?  Thanks

**Comment by nosvalds** [comment]
Should be able to get back to you this week.

**Comment by nosvalds** [comment]
Hi @andylolz  @stevieflow  I've reviewed with the Technical Team and this change is not considered an [Upgrade of the Standard](https://iatistandard.org/en/iati-standard/upgrades/) that needs to follow a full set of change control. We're adding additional metadata into the repo to support the V2 Validator API. 

Keeping a separate copy of this type of information in the Validator repository is something we considered in developing V2. However, as evidenced by maintenance of the current Validator this was a downside requiring a new release of the validator every time a Non-Embedded Codelist was changed for example. Given we are responsible for maintenance of this repository, we find it much easier to make those changes in one place and have them be dynamically used by the Validator. It's also much less confusing for troubleshooting and having the community understand what is the source of truth being used in the validator.

We understand that you have tools using this repo and will give you ample time to review the changes and incorporate them as necessary before merging the PRs. I'm currently targeting to merge this PR (plus v2.02/v2.01) in ~6 weeks (week of 13 September) so we can keep on track with releasing the Public Validator API in September.

I will make sure to update the documentation appropriately, thanks for catching that.

If you want to discuss further please email support@iatistandard.org

**Comment by andylolz** [comment]
@nosvalds Here’s roughly how I’d do this:

 * Create a new github repo called `IATI/validator-rules`. The new repo should contain the `<validation-rules>` that have been added to mapping.xml. This could be in a JSON or XML file.
 * Add github actions on IATI-Codelists and IATI-Codelist-NonEmbedded repos that run on push, and fire a `repository_dispatch` event on the new repo. [Similar to this.](https://github.com/codeforIATI/IATI-Codelists-NonEmbedded/blob/a19d2087/.github/workflows/main.yml#L18-L24) (NB this does break encapsulation, but only a tiny bit)
 * On the new repo, add a github action that runs on `repository_dispatch`. This action should have a workflow matrix across each of the IATI-Codelists version branches ([a bit like this one](https://github.com/codeforIATI/codelist-updater/blob/d59a85dc/.github/workflows/import.yml#L7-L58)). The workflow should:
    * clone the relevant branch of the IATI-Codelists repo, as well as the IATI-Codelist-NonEmbedded repo
    * run a build script that merges mapping.xml with the rules (this is basically the only bit that is new)
    * runs your `python mappings_to_codelist_rules.py` script
    * pushes the generated codelist_rules.json to func-iati-redis-cacher-PROD etc (you’ve already done this bit)

^^ Advantages of doing it this way:

 * It’s decoupled from the SSOT and from the validator, so you maintain encapsulation
 * Users of the SSOT who don’t care about the workings of the validator can ignore it
 * It removes the need to push copies of codelist_rules.json to the various version branches of this repo whenever a non-embedded codelist changes
 * You only have to manage a single copy of the build code, rather than a copy for every version branch. This is easier to manage and removes the possibility of build code on different branches getting out of sync
 * It’s all automated. If a change is pushed to a non-embedded codelist, it magically propagates
 * You don’t need to wait for external people like me to approve it, because it’s almost entirely external to anything we’re using. You can just go ahead and do it

I’d be very happy to have a call to discuss this.

**Comment by andylolz** [comment]
Just bumping this a month on, @nosvalds. Would you be interested in discussing further, e.g. on a call?

**Comment by nosvalds** [comment]
Hi @andylolz, if you want to discuss further please email support@iatistandard.org as requested before. We aren't monitoring these PRs for comments at this time.

**Comment by nosvalds** [comment]
This is now planned for the week of 20 September

**Comment by nosvalds** [comment]
Closing this PR as we will no longer be pursuing merging these changes in and will explore a separate repos option when we have time in 2022.

**Pull request closed without merging**
